### PR TITLE
[RFC] Add vimexpect library and example gdb plugin

### PR DESCRIPTION
**Pull request by tarruda**
This library makes it easier to script communication with interactive programs. It is similar to what the "expect" tcl extension does, but uses an object oriented API and is designed to integrate nicely with Neovim job control.

### CODE REVIEW

**Comment by fwalch** [comment]
What about publishing the plugin "standalone", e.g. in tarruda/nvim-gdb?

**Comment by philix** [comment]
> What about publishing the plugin "standalone", e.g. in tarruda/nvim-gdb?

+1 for that even though having a GDB plugin builtin would be nice.

**Comment by Pyrohh** [comment]
Or maybe just under `neovim/nvim-gdb`?

**Comment by tarruda** [comment]
> What about publishing the plugin "standalone", e.g. in tarruda/nvim-gdb?

I think Neovim still needs one extra feature before a debugger plugin can be polished enough for general use: mappings local to window and/or tabpage(That way the debugger shortcuts can be activated only in the "debugging view").

Plus this one is more a proof of concept, and the name is a bit misleading: It's not a generic gdb plugin for Neovim, but a gdb plugin specifically created for developing Neovim.

File: runtime/autoload/vimexpect.vim [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+" Push a state to the state stack
+function s:Parser.push(state)
+  call add(self._stack, a:state)
+endfunction
+
+
+" Pop a state from the state stack. Fails if there's only one state remaining.
+function s:Parser.pop()
+  if len(self._stack) == 1
+    throw 'State stack cannot be empty'
```
Comment (ZyX-I) [inline runtime/autoload/vimexpect.vim]: I would highly suggest make all thrown messages use “{plugin prefix}:{error code}:{Message}” convention. Reason: thrown errors are normally part of the API and `{Message}` part does not look like good candidate for matching in `catch` (too long, looks much like a subject to translation and is definitely a subject to typos correction or convention switches like `State stack cannot be empty` vs `State stack cannot be empty.`). I.e.

    throw 'vimexpect:emptystack:State stack cannot be empty'

or

    throw 'emptystack:State stack cannot be empty'

matched with `catch /^vimexpect:emptystack:/`.

This is only the case if you want thrown errors be the part of the API. I think this is good for libraries.

Comment (tarruda) [inline runtime/autoload/vimexpect.vim]: :+1:

**Comment by justinmk** [comment]
> mappings local to window and/or tabpage

@tarruda  What's wrong with buffer-local mappings? (`nnoremap <buffer> foo ...`) I was also wondering why `:tnoremap` is needed instead of using buffer-local mappings.

**Comment by tarruda** [comment]
@justinmk Tab or window mappings would simplify the creation of debugger-specific mappings(continue, interrupt, step...) that are only active in the debugging tab.

tnoremap is used to enable the mappings even when the gdbclient terminal is focused

**Comment by ZyX-I** [comment]
@justinmk `tnoremap` is for terminal mode. It is needed because terminal mode has a great majority of differences with insert mode (e.g. you can’t move cursor using `<C-o>h`). Tabpage-local mappings have sense if you understand the idea of tab pages as workspaces with different purpose. Do not know why window-local mappings may be needed.

**Comment by justinmk** [comment]
> Tab or window mappings would simplify the creation of debugger-specific mappings(continue, interrupt, step...) that are only active in the debugging tab

So the idea is that "debug mode" would be enabled in one tab, but not globally. I don't see why global mappings for debug mode would be a problem though (which are removed after debug mode is done). Maybe what we really need is a way for global (and buffer-local) mappings to temporarily mask global/buffer mappings and then restore the original mappings reliably after the mode is exited.

>  Tabpage-local mappings have sense if you understand the idea of tab pages as workspaces with different purpose.

If I load a buffer into such a tabpage, what happens with the buffer's FileType/ftplugin which sets up buffer-local mappings? This is a lot of cognitive overheard for the user.

**Comment by ZyX-I** [comment]
> If I load a buffer into such a tabpage, what happens with the buffer's FileType/ftplugin which sets up buffer-local mappings? This is a lot of cognitive overheard for the user.

Why do you think something should happen? Vim lived with two tables with mappings for years, adding third table should not make any difference. If code was sane enough to just iterate over tables in place of doing “loop unroll” manually that would even be trivial to implement.

I do not think there is much additional overhead to what is *already* there because of buffer-local mappings. I do not see much difference between “filetype is X” condition and “tabpage is Y” (as long as they are indicated properly).

**Comment by tarruda** [comment]
> So the idea is that "debug mode" would be enabled in one tab, but not globally. I don't see why global mappings for debug mode would be a problem though (which are removed after debug mode is done). 

While this is possible, it makes things more complicated and error-prone for plugins which will repeat code for saving/restoring mappings. Not to mention it won't handle more complicated scenarios like this one:

- plugin x opens a tabpage with custom mappings.
- plugin y opens a tabpage with custom mappings while plugin x tabpage is active.
- plugin x closes its tabpage, restoring user mappings.
- plugin y closes its tabpages, but it will restore plugin x mappings(breaks user mappings).

With tab-local mappings this is greatly simplified because everything is handled by the mapping engine automatically.

> Maybe what we really need is a way for global (and buffer-local) mappings to temporarily mask global/buffer mappings and then restore the original mappings reliably after the mode is exited.

Its another way to handle the problem, but still more complicated than tab-local mappings that have a predefined scope.

> If I load a buffer into such a tabpage, what happens with the buffer's FileType/ftplugin which sets up buffer-local mappings? This is a lot of cognitive overheard for the user.

First search buffer, then tab, then global mappings.

**Pull request merged**